### PR TITLE
feat: add support for glob and regex patterns in package name matching

### PR DESCRIPTION
**Pull request by officialasishkumar**
### Description

This pull request improves the package record query functionality in `SparseRepoData` to support glob and regex patterns for package names, alongside exact matches. The implementation now efficiently handles mixed queries and includes comprehensive tests to ensure correct behavior for pattern-based queries.

**Enhanced pattern matching in package queries:**

* Refactored the core logic in `SparseRepoData::load_matching_records` to separate and handle exact package name matches and pattern-based matches (glob, regex, or nameless specs). Exact matches use efficient binary search, while pattern matches iterate over all package names. [[1]](diffhunk://#diff-c294f18d3d0384e772f49101ed172a53b10934efebce51eb39588c237262b12aL350-R431) [[2]](diffhunk://#diff-c294f18d3d0384e772f49101ed172a53b10934efebce51eb39588c237262b12aL364-R448)
* Added support for `PackageNameMatcher` in imports, enabling pattern matching capabilities for package names.

**Testing improvements:**

* Added new tests for glob patterns, regex patterns, and mixed queries to verify correct matching behavior for various query types, including edge cases.

Fixes https://github.com/conda/rattler/issues/2122

### How Has This Been Tested?

i tested this by searching for two regex

```
❯ pixi run -- cargo run -p rattler-bin --release -- search "openssl*" --limit 3
    Finished `release` profile [optimized] target(s) in 1.01s
     Running `target/release/rattler search 'openssl*' --limit 3`
Searching for 'openssl*' on osx-arm64
Channels: https://conda.anaconda.org/conda-forge/
Found 101 matching records in 1.310689584s

openssl (101 versions)
  3.6.1 hd24854e_1 [osx-arm64] https://conda.anaconda.org/conda-forge/
  3.6.1 hd24854e_0 [osx-arm64] https://conda.anaconda.org/conda-forge/
  3.6.0 h5503f6c_1 [osx-arm64] https://conda.anaconda.org/conda-forge/
   ... and 98 more versions

❯ pixi run -- cargo run -p rattler-bin --release -- search "^python-.*$" --limit 3
    Finished `release` profile [optimized] target(s) in 0.85s
     Running `target/release/rattler search '^python-.*$' --limit 3`
Searching for '^python-.*$' on osx-arm64
Channels: https://conda.anaconda.org/conda-forge/
Found 11485 matching records in 318.13175ms

python-a2a (1 version)
  0.5.10 pyhcf101f3_0 [noarch] https://conda.anaconda.org/conda-forge/

python-abi3 (30 versions)
  3.13 hdb156c7_2 [noarch] https://conda.anaconda.org/conda-forge/
  3.13 h9bebd56_2 [noarch] https://conda.anaconda.org/conda-forge/
  3.13 hdb156c7_1 [noarch] https://conda.anaconda.org/conda-forge/
   ... and 27 more versions

python-ace (3 versions)
  0.3.1 py312hd7f1de0_1 [osx-arm64] https://conda.anaconda.org/conda-forge/
  0.3.1 py310ha6c330b_1 [osx-arm64] https://conda.anaconda.org/conda-forge/
  0.3.1 py311h9feb11d_1 [osx-arm64] https://conda.anaconda.org/conda-forge/

... and 305 more packages

```

### AI Disclosure

- [x] This PR contains AI-generated content.
  - [x] I have tested any AI-generated content in my PR.
  - [x] I take responsibility for any AI-generated content in my PR.

Tools: Claude, Codex and a little bit of kimi

### Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added sufficient tests to cover my changes.

### CODE REVIEW

**Comment by baszalmstra** [comment]
Thanks for contributing.

> Tools: {e.g., Claude, Codex, GitHub Copilot, ChatGPT, etc.}

You used them all then? 😅 

Please make sure to fill out the template correctly.

**Comment by officialasishkumar** [comment]
@baszalmstra updated it

File: crates/rattler_repodata_gateway/src/sparse/mod.rs [hunk 1]
```diff
@@ -361,14 +393,60 @@ impl SparseRepoData {
...
+                    .filter(|(matcher, _)| matcher.as_ref().is_none_or(|m| m.matches(&name)))
+                    .map(|(_, spec)| spec)
+                    .collect();
+
+                if matching_specs.is_empty() {
+                    continue;
+                }
+
+                // Load and filter records for this package
+                let mut parsed_records = parse_records(
```
Comment (baszalmstra) [inline crates/rattler_repodata_gateway/src/sparse/mod.rs]: This duplicates parsing. Why not first gather all matching package names in exact_specs?

File: crates/rattler_repodata_gateway/src/sparse/mod.rs [hunk 2]
```diff
@@ -1459,4 +1538,134 @@ mod test {
...
+
+        insta::assert_snapshot!(records.join("\n"), @r###"
+        bar
+        baz
+        bors
+        "###);
+    }
+
+    #[test]
+    fn test_mixed_exact_and_pattern_query() {
```
Comment (baszalmstra) [inline crates/rattler_repodata_gateway/src/sparse/mod.rs:1573]: Maybe you can reduce the code duplication in these tests?

Comment (officialasishkumar) [reply crates/rattler_repodata_gateway/src/sparse/mod.rs:1573]: done

File: crates/rattler_repodata_gateway/src/sparse/mod.rs (hunk 1 above)
Comment (officialasishkumar) [reply crates/rattler_repodata_gateway/src/sparse/mod.rs]: done
